lib/Backend/GlobOpt.cpp

```diff
@@ -9230,47 +9243,77 @@ GlobOpt::OptConstFoldBranch(IR::Instr *instr, Value *src1Val, Value*src2Val, Val
     case Js::OpCode::BrNotNeq_A:
         if (!src1Var || !src2Var)
         {
-            return false;
+            ValueInfo *src1ValInfo = src1Val->GetValueInfo();
+            ValueInfo *src2ValInfo = src2Val->GetValueInfo();
+            if (
+                (src1ValInfo->IsNumber() && src1Var && src2ValInfo->IsBoolean() && src1Var != Js::TaggedInt::ToVarUnchecked(0) && src1Var != Js::TaggedInt::ToVarUnchecked(1)) ||
+                (src2ValInfo->IsNumber() && src2Var && src1ValInfo->IsBoolean() && src2Var != Js::TaggedInt::ToVarUnchecked(0) && src2Var != Js::TaggedInt::ToVarUnchecked(1))
+               )
+            {
+                    result = false;
+            }
+            else
+            {
+                return false;
+            }
         }
-        if (func->IsOOPJIT() || !CONFIG_FLAG(OOPJITMissingOpts))
+        else
         {
-            // TODO: OOP JIT, const folding
-            return false;
+            if (func->IsOOPJIT() || !CONFIG_FLAG(OOPJITMissingOpts))
+            {
+                // TODO: OOP JIT, const folding
+                return false;
+            }
+            result = Js::JavascriptOperators::Equal(src1Var, src2Var, this->func->GetScriptContext());
         }
-        result = Js::JavascriptOperators::Equal(src1Var, src2Var, this->func->GetScriptContext());
         break;
     case Js::OpCode::BrNeq_A:
     case Js::OpCode::BrNotEq_A:
         if (!src1Var || !src2Var)
         {
-            return false;
+            ValueInfo *src1ValInfo = src1Val->GetValueInfo();
+            ValueInfo *src2ValInfo = src2Val->GetValueInfo();
+            if (
+                (src1ValInfo->IsNumber() && src1Var && src2ValInfo->IsBoolean() && src1Var != Js::TaggedInt::ToVarUnchecked(0) && src1Var != Js::TaggedInt::ToVarUnchecked(1)) ||
+                (src2ValInfo->IsNumber() && src2Var && src1ValInfo->IsBoolean() && src2Var != Js::TaggedInt::ToVarUnchecked(0) && src2Var != Js::TaggedInt::ToVarUnchecked(1))
```

Comment:
nit: I would refactor this condition (and the one above for BrEq_A) into a method.